### PR TITLE
new diagnosands handler

### PR DESCRIPTION
**Pull request by graemeblair**
Big change to how `declare_diagnosands` works with a new handler. By default, `diagnose_design` still uses the built-in, fast `default_diagnosands` function. However, if you want to roll your own, the new handler allows you to mix-and-match between the defaults and your own. You can just use all the defaults, you can `select = c(rmse, bias)` some to keep or `subtract = type_s_rate` some to remove. You can add your own via the `...` i.e. `declare_diagnosands(median_bias = median(est - estimand)`.

Closes https://github.com/DeclareDesign/DeclareDesign/issues/194

### CODE REVIEW

File: R/declare_diagnosands.R [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                               subset = NULL, 
+                               alpha = 0.05,
+                               label) {
+  
+  options <- quos(...)
+  if (length(options) > 0 && names(options)[1] == "") names(options)[1] <- label
+
+  # subsetting the data -----------------------------------------------------
+
+  subset <- substitute(subset)
```
Comment (nfultz) [inline R/declare_diagnosands.R]: use rlang not substitute

File: R/declare_diagnosands.R [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+      sd_estimate = sd(est),
+      mean_se = mean(se),
+      type_s_rate = mean((sign(est) != sign(estimand))[ p < alpha ] ),
+      mean_estimand = mean(estimand)
+    )
+  
+  if (!missing(select)) {
+    select_quo <- enquo(select)
+    if (quo_is_call(select_quo)) {
+      select_set <- as.character(call_args(select_quo))
```
Comment (nfultz) [inline R/declare_diagnosands.R]: use the existing helper functions for converting to character.

File: R/declare_diagnosands.R [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    if (!all(select_set %in% names(defaults_quos))) {
+      stop("Some of your select set are not included in default diagnosands: ", paste(select_set[!select_set %in% names(defaults_quos)], collapse = ", "), ".")
+    }
+    defaults_quos <- defaults_quos[select_set]
+  }
+  
+  if (!missing(subtract)) {
+    subtract_quo <- enquo(subtract)
+    if (quo_is_call(subtract_quo)) {
+      subtract_set <- as.character(call_args(subtract_quo))
```
Comment (nfultz) [inline R/declare_diagnosands.R]: same

File: R/declare_diagnosands.R [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    } else {
+      subtract_set <- quo_text(enquo(subtract_quo))
+    }
+    if (!all(subtract_set %in% names(defaults_quos))) {
+      stop("Some of your subtract set are not included in default diagnosands: ", paste(subtract_set[!subtract_set %in% names(defaults_quos)], collapse = ", "), ".")
```
Comment (nfultz) [inline R/declare_diagnosands.R]: should be a declare time warning, move to validation function.

File: R/declare_diagnosands.R [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    }
+    defaults_quos <- defaults_quos[!names(defaults_quos) %in% subtract_set]
```
Comment (nfultz) [inline R/declare_diagnosands.R]: setdiff

File: R/declare_diagnosands.R [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+  }
+  
+  if (keep_defaults) {
+    options <- c(options, defaults_quos[!names(defaults_quos) %in% names(options)])
+  }
+  
+  if (length(options) == 0) {
+    stop("No diagnosands were declared.", call. = FALSE)
```
Comment (nfultz) [inline R/declare_diagnosands.R]: declare time error

**Comment by nfultz** [comment]
Most of the logic should get moved from execution time to declare time, eg move from the handler to the handler validation function.

**Comment by graemeblair** [comment]
That's fair; will implement. Thanks.

File: R/declare_diagnosands.R (hunk 2 above)
Comment (graemeblair) [reply R/declare_diagnosands.R]: ooh, thanks, which are those?

**Review by acoppock — APPROVED** [review]
wow lgtm

**Pull request merged**